src/controller/scan/base_controller.go

```diff
@@ -233,78 +266,116 @@ func (bc *basicController) Scan(ctx context.Context, artifact *ar.Artifact, opti
 	return nil
 }
 
-func (bc *basicController) makeReportPlaceholder(ctx context.Context, r *scanner.Registration, art *ar.Artifact, options ...Option) (string, []string, error) {
-	trackID, err := bc.uuid()
+func (bc *basicController) ScanAll(ctx context.Context, trigger string, async bool) (int64, error) {
+	extraAttrs := map[string]interface{}{}
+	executionID, err := bc.execMgr.Create(ctx, job.ImageScanAllJob, 0, trigger, extraAttrs)
 	if err != nil {
-		return "", nil, errors.Wrap(err, "scan controller: scan")
+		return 0, err
 	}
 
-	// Parse options
-	ops, err := parseOptions(options...)
-	if err != nil {
-		return "", nil, errors.Wrap(err, "scan controller: scan")
+	if async {
+		go bc.startScanAll(bc.makeCtx(), executionID, async)
+	} else {
+		if err := bc.startScanAll(ctx, executionID, async); err != nil {
+			return 0, err
+		}
 	}
 
-	create := func(ctx context.Context, digest, registrationUUID, mimeType, trackID string, status job.Status) error {
-		reportPlaceholder := &scan.Report{
-			Digest:           digest,
-			RegistrationUUID: registrationUUID,
-			Status:           status.String(),
-			StatusCode:       status.Code(),
-			TrackID:          trackID,
-			MimeType:         mimeType,
-		}
-		// Set requester if it is specified
-		if len(ops.Requester) > 0 {
-			reportPlaceholder.Requester = ops.Requester
-		} else {
-			// Use the trackID as the requester
-			reportPlaceholder.Requester = trackID
-		}
+	return executionID, nil
+}
 
-		_, e := bc.manager.Create(reportPlaceholder)
-		return e
+func (bc *basicController) startScanAll(ctx context.Context, executionID int64, async bool) error {
+	if async {
+		// if async, this method is running in another goroutine ensure the execution exists in db
+		err := lib.RetryUntil(func() error {
+			_, err := bc.execMgr.Get(ctx, executionID)
+			return err
+		})
+		if err != nil {
+			log.Errorf("failed to get the execution %d for the scan all", executionID)
+			return err
+		}
 	}
 
-	if hasCapability(r, art) {
-		var producesMimes []string
-
-		for _, pm := range r.GetProducesMimeTypes(art.ManifestMediaType) {
-			if err = create(ctx, art.Digest, r.UUID, pm, trackID, job.PendingStatus); err != nil {
-				return "", nil, err
-			}
+	batchSize := 50
+	for artifact := range ar.Iterator(ctx, batchSize, nil, nil) {
+		scan := func(ctx context.Context) error {
+			return bc.Scan(ctx, artifact, WithExecutionID(executionID))
+		}
 
-			producesMimes = append(producesMimes, pm)
+		if err := orm.WithTransaction(scan)(ctx); err != nil {
+			// Just logged
+			log.Errorf("failed to scan artifact %s, error %v", artifact, err)
 		}
+	}
 
-		if len(producesMimes) > 0 {
-			return trackID, producesMimes, nil
+	count, err := bc.taskMgr.Count(ctx, q.New(q.KeyWords{"execution_id": executionID}))
+	if err != nil {
+		log.Errorf("failed to get the count of tasks for scan all, error %v", err)
+		return err
+	} else if count == 0 {
```

Comment:
If all callings of `bc.Scan()` failed(no task record is created), the task count is 0, but this doesn't the case of `no task found`

Reply:
Mark the execution to be done here is ok.

The scan method returns the error only when the artifact is not supported by the scanner or the artifact is scanning by the scanner.

What is the status of the execution when 100 artifacts found and 99 of them scan failed and 1 success? According to the status change handler, is the status of the execution success?

Reply:
If the execution contains failed tasks, it is marked as failed